Refactor MacOS and provide a coroutine-driven runloop

willglynn opened the pull request:
This is a major MacOS refactoring, per discussion in #219, and among other things it closes #219.

## Summary

This PR refactors the MacOS event loop such that it occurs in a `Runloop`. `EventsLoop` calls `Runloop::work(timeout: Timeout)`, which performs a cycle of the event loop (blocking for up to the specified timeout), delivering zero or more events to a queue, and returns.

This PR also adds a second `Runloop` implementation, available via `--features context`, which runs in a coroutine. MacOS X event handling can enter inner loops (see [this comment](https://github.com/tomaka/winit/issues/219#issuecomment-315830359) in particular and #219 in general), and the only way to connect `winit`'s polling-centric API with the MacOS event loop while preserving the contracts of both is to run the MacOS event loop in a coroutine.

## Comparative behavior

Consider a user with code like this:

```rust
loop {
    events_loop.poll_events(|event| {
        // handle event
    });
    
    // draw frame
}
```

Now let's look at what happens on MacOS during a window resize. During a resize, `forward_event_to_cocoa()` enters an internal loop where Cocoa callbacks can generate `Resized` events, but where `forward_event_to_cocoa()` can not return.

In `master`, `poll_events()` stores the user's callback in a place where Cocoa callbacks can reach it. This lets the `Resized` events get delivered immediately, but the user can't _do_ anything with them, since `poll_events()` continues to block until the user stops resizing the window. No frames get drawn.

In this PR without `--features context`, the Cocoa callback creates `Resized` events and puts them in the queue immediately, but they can't be delivered because the `forward_event_to_cocoa()` is stuck in an internal loop. This means the user's code does not see the events immediately, but the net effect is the same -- `poll_events()` blocks for exactly the same amount of time, and the next frame is delayed exactly as long.

In this PR with `--features context`, the `Resized` events land in the `Shared` event queue immediately, and the runloop coroutine yields immediately, even though it's now inside another event loop. `poll_events()` returns immediately, the user can draw a frame, and the next call to `poll_events()` resumes that interior loop where it left off.

Likewise, in `master` and in this PR without `--features context`, `Proxy::wakeup()` events are deferred until after a resize. In this PR with `--features context`, wakeup events are delivered promptly from any state.

## Structure

The basic `Runloop` does everything in `Runloop`. The `context`-enabled version moves those functions to `InnerRunloop`, and adds a `Runloop` with the same public interface whose purpose is to start an `InnerRunloop` coroutine and context switch into it as needed.

## Entering the coroutine

After initialization in `Runloop::new()`, `Runloop::work()` is the only place where the main thread context switches into the `InnerRunloop` coroutine.

`Runloop::work()` is called only by `EventLoop::get_event()`. Whatever invariants about the main thread are true at that point remain true for the entire duration of the inner runloop. For example, we know that the main thread is not holding locks inside `Shared`, so the coroutine can acquire and release locks on `Shared` without deadlocking on the main thread.

`Runloop::work()` checks the `NSThread`'s identity to ensure that the coroutine can only be resumed from the main thread.

## Moving data into the coroutine

The initial call into the coroutine entrypoint needs to bring an `InnerRunloop`, and all subsequent calls into the coroutine bring a `Timeout`. I made this happen by combining three properties:

* `context` can carry a `data: usize` along during a context switch.
* When execution is transferred into a coroutine, the caller stops running until execution transfers back.
* `Option::take()` moves a value out of the `Option`.

Put together, this means the caller can declare a local `Option`, pass a `&mut` of it into the coroutine, and the coroutine can safely `.take()` its value. Again, there's no concurrency at work -- everything executes sequentially -- so we can guarantee that there's only one mutable borrow to the caller's `Option`.

Actually doing this via a `usize` requires `&mut` as `*mut` as `usize` on the way down, and `usize` as `*mut` transmute `&mut` on the way up. One could transmute straight from `usize` to `&mut`, but I kept `*mut` for symmetry.

Calls into the coroutine look like:

```rust
// caller
let mut input: Option<Foo> = Some(Foo);
context.resume(&mut input as *mut Option<Foo> as usize);

                // coroutine's resume() returns, holding the &mut Option
                let t: context::Transfer = context.resume(…);
                let input: *mut Option<Foo> = t.data as *mut Option<Foo>;
                let input: &mut Option<Foo> = mem::transmute(input);
                let input: Foo = input.take().unwrap();
                // input is now moved to the coroutine
                // coroutine eventually returns to the caller
                t.context.resume(…);

// caller's context.resume() returns
// input = None, since the value was taken by the coroutine
```

## Inside the coroutine

`yield_to_caller()` is the place where the coroutine context switches back to `Runloop::work()`, and it is therefore also the place where the coroutine resumes.

`yield_to_caller()` sets a thread local cell containing the caller's context when execution switched into the coroutine, and it moves the context out of that cell before it switches back. If that cell is full, then we are currently inside the coroutine; if that cell is empty, then we are not.

`yield_to_caller()` also sets a thread local cell containing the caller's `Timeout`, which can be retrieved by `fn current_timeout() -> Option<Timeout>`. The inner runloop uses this when asking Cocoa to receive an event.

The coroutine's `InnerRunloop` looks very much like the normal blocking `Runloop`. It tries to receive an event from Cocoa, forwards it back to Cocoa, translates it into zero or more `Event`s, and posts them to the queue.

## Exiting the coroutine

`Shared::enqueue_event()` enqueues the event and then tries to wake the runloop, and the coroutine version of `Runloop:::wake()` calls `yield_to_caller()`. This means that if we enqueue an event from inside the coroutine -- for example, from the normal inner runloop or because a Cocoa callback posted an event -- then execution immediately returns to `EventLoop::get_event()`, which checks `Shared`'s event queue, finds an event, and returns it to its caller.

If `Runloop::wake()` finds that its caller is _not_ inside the coroutine -- for example, because it's on a different thread calling `Proxy::wakeup()` -- it calls `CFRunLoopWakeUp(CFRunLoopGetMain())` in hopes of stirring the main runloop. The main `CFRunLoop` inside the coroutine wakes up, regardless of whether that's the normal loop or whether it's that an inner loop caused by a resize. This wakeup triggers a `CFRunLoopObserver`, which calls `yield_to_caller()`. That gets us back to the main thread even if we're stuck inside someone else's loop.

In addition to the runloop observer, the coroutine has a timer whose callback calls `yield_to_caller()`. This causes needless context switches, but it also sets an upper bound on how long the coroutine can run continuously without yielding, which significantly improves worst-case latency.

jwilm commented:
I'm really excited about the work you're doing here. The stack management on winit's side seems good, and not running user code in the coroutine is a good choice. One nagging concern I've had is how safe it is to suspend execution of the Cocoa APIs? For instance, what happens when there are multiple windows being managed? It seems that both windows could be suspended in a coroutine at the same time. What happens if there's shared state in the Cocoa side?

One situation where this could show up is with a modal dialog. The main window may have already processed some events and suspended the Cocoa event loop, opened the modal window, and started processing events there.

By the way, my question is coming from a place of ignorance about how Cocoa works in general, and this may just be a non-issue.

willglynn commented:
> One nagging concern I've had is how safe it is to suspend execution of the Cocoa APIs? For instance, what happens when there are multiple windows being managed? It seems that both windows could be suspended in a coroutine at the same time. What happens if there's shared state in the Cocoa side?

To start: it's important that there's no preemption of Cocoa. The `Runloop` coroutine can yield back to the user from the deepest depths of a Cocoa call stack, but even in that case, it's returning to the user from within a Cocoa callback. In other words, writing this code with `--features context`:

```rust
loop {
    events_loop.poll_events(|event| {
        // handle event
    });
    
    draw_frame();
}
```

…lets the user `draw_frame()` at the same point as the user can right now:

```rust
events_loop.run_forever(|event| {
    // handle event
    if it_is_time_to_draw_a_frame {
      draw_frame();
    }
});
```

Both allow the user to interact with Cocoa from the main thread in the middle of a Cocoa callback. Cocoa's state is identical in both cases; the coroutine approach just swizzles around the stack pointer so that the user can do this outside`winit` instead of doing it from a callback inside `winit`.

As for multiple windows, `examples/multiwindow.rs` uses a single event loop (and thus a single `Runloop` of either kind) to serve multiple windows. That's the normal, expected, and sensible way to do things. Working that single loop will provide events for all the windows.

I'm not sure what would happen if you tried creating multiple windows with unrelated event loops, and you're correct, that could give you multiple coroutines suspended at different points. However, given that Cocoa actually provides only one event loop, the `NSThread` check won't switch into _any_ runloop from the wrong thread, and the runloops only yield at defined points, I don't think there's any UB potential.

I think the worst case is that you could construct something like:

```rust
events_loop_1.run_forever(|event| {
    events_loop_2.run_forever(|event| {
        // ¯\_(ツ)_/¯
    });
});
```

The new runloops follow the same pattern as the old ones, so such a construction should succeed or fail in much the same way that it would now.

mitchmindtree commented:
Just wanted to mention that I'm also quite excited for this PR! That said, I haven't had a close look yet - I'll try to get around to testing and reviewing this soon.

willglynn commented:
FYI: I will have intermittent internet connectivity starting tomorrow for about a week. That's partly I pushed to write all this up and open the PR today. I expect that proper review of this changeset will take time; I wanted to get that process started before I left, and I wanted to provide enough information that it could proceed without requiring me to be immediately available.

Having said that, I will still have a computer with a Rust compiler and a `winit` checkout :-) If you have a question, or want a change, or find a problem, let me know and I'll do what I can.

willglynn commented:
I found and fixed a regression versus `master` without `--features context`. Specifically, `EventsLoopProxy::wakeup()` + `run_forever()` would not deliver `Awakened` events until some other event arrived.

The old approach was for `EventsLoopProxy::wakeup()` to post an `NSEvent` of type `NSApplicationDefined` to an OS event queue, which was translated into an `Event::Awakened` on receipt. The new approach in this PR is for `EventsLoopProxy::wakeup()` to post an `Event::Awakened` to a `winit` event queue, and for posting events to imply calling `Runloop::wake()`, which is supposed to do _something_ to reduce the time spent between "event is in the queue" and "user gets the event".

The coroutine runloop implementation of `Runloop::wake()` causes a context switch. If we're trying to wake from inside the coroutine, it can context switch back directly. Otherwise, `Runloop::wake()` assumes the runloop is asleep on a different thread, so it tells the OS to wake the runloop. Waking the runloop triggers the runloop observer, and the runloop observer callback causes a context switch. The net effect in both cases is that the main thread checks the event queue and returns to the user.

The blocking runloop tries to wake itself by asking the OS, but the runloop was missing logic to check for queued events, which meant it would sit in `receive_event_from_cocoa()` until an event got delivered. This is fine when called via `poll_events()`, but `run_forever()` blocks forever, so the next check could be arbitrarily far in the future.

I addressed this by making the non-coroutine `Runloop::wake()` post an `NSApplicationDefined` event just like before. Unlike before, this event does nothing -- it does _not_ translate into an `Awakened` event, since that event is already in the queue. Its function is to allow `receive_event_from_cocoa()` to return so that the runloop can check `Shared::has_queued_events()` and return as needed.

---

All this prompted me to write a tool to measure wakeup latency, which I stuck in `examples/wakeup_latency.rs`. It spawns a thread to call `EventsLoopProxy::wakeup()` 60 times per second, uses `run_forever()` in the main thread to receive `winit::Event::Awakened` 60 times per second, and calculates how long the wakeup took to arrive. Every 300 events (=~ 5 seconds), it calculates the mean along with various percentiles and prints them.

### Typical wakeup latency

Baseline from current `master`:

```
wakeup() -> Event::Awakened latency (all times in µs)
mean    max     99%     95%     50%     5%      1%      min
1219    90248   35209   434     330     148     115     102
312     11997   441     427     310     100     78      64
288     600     448     428     291     153     137     124
308     842     596     433     317     184     143     108
301     446     436     430     321     151     131     127
329     445     441     433     334     209     171     135
329     440     438     432     333     217     162     130
327     465     441     432     332     199     155     129
313     446     439     432     323     186     142     97
311     1272    441     429     320     167     132     118
259     1024    519     428     231     138     113     92
^C
```

This PR without `--features context` now fares comparably:

```
wakeup() -> Event::Awakened latency (all times in µs)
mean	max	99%	95%	50%	5%	1%	min
902	72556	15382	406	301	140	112	78
301	429	413	404	316	154	137	128
314	505	415	405	319	204	183	144
297	418	413	405	315	153	136	130
266	417	409	403	246	147	127	106
318	422	411	403	321	208	199	196
313	455	415	405	321	197	150	137
298	5631	414	405	294	159	114	87
267	1549	409	403	227	176	137	96
305	630	418	406	316	194	144	79
^C
```

This PR with `--features context` runs significantly faster, with median latencies lower than the minimum latencies previously seen:

```
wakeup() -> Event::Awakened latency (all times in µs)
mean	max	99%	95%	50%	5%	1%	min
491	40276	20302	159	51	17	4	2
67	178	164	156	57	17	5	2
82	207	199	160	76	25	13	8
80	234	199	159	75	23	12	3
79	228	180	161	70	25	19	5
74	216	171	158	64	25	20	5
75	194	183	159	71	24	11	3
74	199	173	158	72	25	18	9
68	199	160	154	64	20	10	4
65	208	168	154	62	22	13	3
^C
```

### Resizing wakeup latency

Here I tried to grab the window for resizing the moment it printed the first summary line, then hold it for ~6 seconds.

Current `master` shows extreme latencies, since the `NSApplicationDefined` events (which translate into `Awakened` events) do not get delivered until the resize is complete:

```
wakeup() -> Event::Awakened latency (all times in µs)
mean	max	99%	95%	50%	5%	1%	min
1499	113548	58438	212	107	72	65	63
3127967	6012464	5959120	5751020	3154140	181	128	104
31976	586795	530781	310085	165	137	112	99
^C
```

This PR without `--features context` fares no better, since like `master` its runloop gets stuck in `forward_event_to_cocoa()`:

```
wakeup() -> Event::Awakened latency (all times in µs)
mean	max	99%	95%	50%	5%	1%	min
1157	87584	34583	406	303	137	112	82
2931622	5919778	5860272	5640558	3056384	204	188	140
137012	1219426	1165654	951234	304	135	110	67
^C
```

This PR with `--features context` suffers versus the not-resizing case, since the main thread still becomes busy doing other things during a resize, but it actually does deliver `Awakened` events promptly:

```
wakeup() -> Event::Awakened latency (all times in µs)
mean	max	99%	95%	50%	5%	1%	min
332	32771	6514	130	37	11	3	2
876	30898	28488	232	38	10	3	3
41	158	97	78	44	6	3	3
^C
```

willglynn commented:
I rebased this PR (dff60d32 => 6ea5d452).

willglynn commented:
> I wonder if we could even include the more detailed overview (from the top comment of this PR) at the top of the module? I think it provides a valuable overview of the reasoning and workings behind the context switching which future contributors could benefit from.

Sure, I'll work it in.

> One regression that I noticed when running the examples with `--features "context"` is that the CPU usage no longer drops to 0% when they are completely unfocused and idle - instead they seem to hover at about 4%. @willglynn do you have any idea what might be causing this?

That's almost certainly the timer. I added it to contain the worst case latency, but there's no reason at all for it to run when the application is backgrounded. It's _probably_ enough to activate the timer only during a resize, bounded by `windowWillStartLiveResize`/`windowDidEndLiveResize` callbacks on `NSWindowDelegate`, though perhaps it would be better to use it as an alarm clock to unconditionally wake us out of an event send.

I also had the thought that the runloop observer could possibly be replaced with `CFRunLoopPerformBlock`. Instead of context switching each time the runloop sleeps/wakes _in case_ the user wants their thread back, we can enqueue a block that performs a context switch only when the user actually indicates they want their thread via `wakeup()`.

My previous measurements were subjective and visual, which was sufficient to address the problems I observed in `master` but is not a great basis for making efficient tradeoffs. Now that `--example wakeup_latency` spits out concrete numbers, I'll run some experiments and see what I can do.

willglynn commented:
I rebased again (6ea5d45 => 8c2aaeeba0ca31d72647ce1ffe4abaafc34a69b4), this time fixing conflicts from #229.

willglynn commented:
Trading out the unconditional runloop observer for an as-needed queued block shows worse steady state timings, but they're still better than `master`, resizing still works fine, and it avoids some spurious callbacks:

```
wakeup() -> Event::Awakened latency (all times in µs)
mean	max	99%	95%	50%	5%	1%	min
521	44335	12039	189	82	24	12	5
78	260	196	163	74	20	8	5
89	223	194	150	81	36	24	16
86	225	190	146	81	27	12	2
94	266	207	155	83	40	24	6
89	256	198	159	81	43	24	4
87	196	187	147	79	34	23	15
86	215	184	139	77	43	27	15
80	249	183	142	74	30	12	7
^C
```

I also found that [`core-foundation-sys` doesn't provide `CFRunLoopPerformBlock`](https://github.com/servo/core-foundation-rs/blob/master/core-foundation-sys/src/runloop.rs#L122) so I pulled in the [`dispatch` crate](https://docs.rs/dispatch/0.1.2/dispatch/) to say something like `dispatch_async_f(dispatch_get_main_queue(), …)` instead. Point is, another thread can somehow gets the main thread to call `yield_to_caller()` on demand directly.

willglynn commented:
I ripped out the unconditional timer, and added a check -- if you enter the coroutine with `Timeout::Now`, it'll use `dispatch_after()` to enqueues a block for 2 ms after entering the runloop.

This puts a slightly lower bound on the amount of time the runloop can block the main thread -- 2ms from runloop entry vs 2.5ms average waiting for a 5ms timer -- and it's consistent to boot. It also does nothing in the `Timeout::Forever` case, so the examples' CPU usage stats are better.

Coupled with the previous change, this cuts Activity Viewer's "idle wakeups" on `--example wakeup_latency` from ~250/s (the timer, runloop observer, and the helper thread sending wakeups) to ~60/s (just the helper thread).

willglynn commented on src/platform/macos/events_loop/runloop_context.rs:
```diff
@@ -0,0 +1,276 @@
+
+    // Turn this into a mutable borrow, then move the inner runloop into the coroutine's stack
+    let mut inner: InnerRunloop =
+        unsafe { mem::transmute::<*mut Option<_>, &mut Option<_>>(inner) }
```
I chose to see this as a chain of `&mut` -> `*mut` -> `usize` -> `*mut` -> `&mut` casts. The thread always goes straight down that line, executing everything top to bottom, and I opted to reinforce that idea by fearlessly reaching for `transmute`.

`transmute` is a bold statement, but so is "these things are guaranteed to happen in this order, so I know for a fact that this `usize` is actually a `&mut Option`". My opinion was that this particular construction improves clarity here, especially when the alternative is `as_mut() -> Option<&mut Option<T>>` in a confusing context where we've already got `&mut`s and `Option`s in play.

Having said all that, if you want me to change it, let me know :-)

willglynn commented:
I rebased again (3f99dd80 => 0bb78111).

trishume commented on src/platform/macos/events_loop/runloop_context.rs:
```diff
@@ -0,0 +1,359 @@
+        let (most_recent_invocation, _) = current_counter.overflowing_sub(1);
+
+        // Are we the most recent call?
+        if most_recent_invocation == this_invocation {
```
I don't like how if you run the event loop with `Timout::Now` 1,000 times in quick succession, say because there are no events to read, it will queue up 1000 blocks. Then 999 of the blocks will fail this check and do nothing. Timer coalescing will at least make there be less than 1000 wakeups, but it's still a bunch of extra work and memory. Or am I misunderstanding how this works?

Ideally there should only be one block in the queue at any time. You could do this with an `in_flight : AtomicBool` variable where you compare and swap `true` with a check of `false` before queueing a block, and then in the block you set the variable to `false` before yielding.

willglynn replied:
Hmm. Thinking aloud in concrete terms:

Let's say we enqueue a block for 5 ms from now, return, and re-enter 2 ms later wanting another 5 ms timer. `dispatch_async_f()` does not provide a means to cancel a queued block, so at the moment we re-enter, our status is that we have a block queued for 3 ms from now for the previous cycle and we have no block queued for 5 ms for this cycle. 

The current code will queue another block and ignore any firings from the previous block(s). This wastes some RAM and some CPU cycles, but if you're polling events in a tight loop the way you described, it's not causing extra wakeups or context switches because you're never letting it sleep anyway. Does this overhead affect real programs? ¯\\\_(ツ)\_/¯

That said, we could reduce this overhead by scheduling blocks less precisely. This is a fallback mechanism to reduce worst-case latency, so there's not any kind of strict timeout policy, it just needs to get control reasonably soon even if the coroutine gets stuck somewhere inside a Cocoa resize handler.

We can waste fewer resources by keeping two bits of state: one indicating the current deadline (which we update unconditionally each time we enter with `Timeout::Now`), another indicating if there is presently a block queued. In this example, the second cycle would reset the deadline to 5 ms from now, but it would see there's a block queued and avoid scheduling another one. When the block gets invoked, it checks the deadline: deadlines in the past `yield_to_caller()`, and deadlines in the future prompt the block to schedule a followup. If the block in this example was invoked on time, it would schedule another for 2 ms from now; if it was invoked late, it'll yield immediately.

I'll hook that up and update the PR.

willglynn replied:
It's not _exactly_ what I described here, but cb5df71 now avoids scheduling overlapping blocks. There is always at most one queued block, and depending on circumstances, that block can decide to a) yield, b) queue another block for later, or c) do nothing.

willglynn commented:
Coroutine support landed in Rust a few hours ago (rust-lang/rfcs#2033), which adds a new `std::ops::Generator` trait.

This PR is structured in a way that would permit a `runloop_generator.rs` implementation to live alongside `runloop_context.rs` and `runloop.rs`. However, this changeset is already too large. Instead of making it larger, I was planning to worry about adding a nightly-only implementation as a separate PR once this one lands.

Diggsey commented:
What's this blocked by?

tomaka commented:
I guess we could merge this after conflicts are resolved.

Diggsey commented:
I have a branch with the conflicts resolved here: https://github.com/tomaka/winit/compare/master...Diggsey:coroutines

Diggsey commented:
I wonder: should the `coroutine` feature be enabled by default? If not, it will regress the behaviour on osx by default.
So far this PR does not expose the new API surface: there will need to be another PR to add "get event with timeout" functionality to the other platforms, and also it would be good to get back the iterator version which used to exist.

tomaka commented:
That's the thing I was hesitating on with this PR. I don't like having two different implementations of the same API, because the one not enabled by default will likely have undetected bugs.

Diggsey commented:
The problem with not accepting it, is that the existing "fix" (not exposing this kind of API) didn't really fix the problem: it just pushed dealing with the issue onto the consumers of winit. In practice this has resulted in an overall worse experience on *all* platforms, because frankly it's a massive PITA to deal with, and there are some issues I'm not even aware of a solution to.

What about simply making `coroutine` a default feature? Clients will still be able to disable it if they don't care about the resizing behaviour on osx.

willglynn commented:
@tomaka I think there will be more than two implementations, which is why I refactored everything in favor of [minimizing the feature-dependent code](https://github.com/willglynn/winit/blob/cb5df718d5583e3ef459e79fefe5c37967f55564/src/platform/macos/events_loop/runloop.rs). I chose `context` for a second implementation because it let me fix #219 in stable Rust. Judging by upstream activity, I expect we'll be eventually able to write a third implementation in nightly Rust, which will eventually make its way into stable Rust.

Implementation | Resizing works | Stable Rust | Extra dependencies
--- | --- | --- | ---
`runloop.rs` | ❌ | ✅ | None
`runloop_context.rs` | ✅ | ✅ | `context`
Native Rust coroutines | ✅ | ❌ => ✅ | None

Until it's possible to do the same thing in stable Rust – on a version of stable Rust that you're willing to require – I think it's appropriate to have multiple implementations despite the cost.

Diggsey commented:
I'd really like to see this get merged 😉
@willglynn does my resolution of the merge conflicts look correct to you? If so, could you update your PR?

bzm3r commented:
Where are we on this?

willglynn commented:
As far as I know, there's no outstanding changes needed, this just… never got merged. I haven't been following along with `winit/master` changes since roughly cb5df718d5583e3ef459e79fefe5c37967f55564. If all this needs is another rebase, I'll take another look, but I'm not eager to invest more time into this PR if it's just going to sit here again.

Diggsey commented:
@willglynn it was waiting on you to review (and apply) my resolution of the merge conflicts on this PR: https://github.com/tomaka/winit/compare/master...Diggsey:coroutines

Unfortunately there are new merge conflicts since then. I will resolve them and update my branch.

willglynn commented:
@Diggsey: Sure, I know this PR has conflicts now. I haven't reviewed either `winit/master` for context or your fork of this PR for correctness. My status comment is more general: this PR _was_ mergeable for quite a while, there were zero outstanding issues that I knew about, and the PR was neither accepted nor rejected.

I know about one outstanding issue now (merge conflicts), and if that's all that needs to happen to get this merged, I'm happy to fix it. On the other hand, @tomaka [recently expressed hesitation](https://github.com/tomaka/winit/pull/237#issuecomment-347472374) about this PR, and viewed in light of [his general aversion to this concept in #219](https://github.com/tomaka/winit/issues/219#issuecomment-315832499), I'm not inclined to hop onto the rebase treadmill without concrete guidance from the project about what's needed to make this happen.

Diggsey commented:
@willglynn well tomaka said:
> I guess we could merge this after conflicts are resolved.

I've resolved the merge conflicts on my fork: I guess we could merge this after conflicts are resolved.

jwilm commented:
We are really excited to have this for Alacritty. @tomaka, do you feel your concerns have been adequately addressed?

tomaka commented:
> @tomaka, do you feel your concerns have been adequately addressed?

I'd be up for merging if the conflicts are solved.

jrmuizel commented:
@willglynn, given @tomaka's willingness to merge do you think you'll be able to do a rebase soon?

tomaka commented:
The problem is that the OSX backend has received some behaviour changes since then. I'm not sure if it's a good idea to have two different implementations of the same backend.

francesca64 commented:
@willglynn I'm trying to understand how to proceed with this, and I'm a bit confused on why there needs to be two implementations. What reason would there be for not using the coroutine feature?

francesca64 commented:
>it seems a lot like we should be looking at changing winit's API instead, in which case we wouldn't need to do any of this.

That's something I've been actively investigating lately; I'm increasingly getting the impression that we'll inevitably have to invert the control flow. I just need to make sure I understand things well enough to be able to make a strong case for it.

It seems that broadly speaking, such an API change would lead to the backends becoming more idiomatic (wrt the design of the underlying platform APIs) and less complex. Even without considering functionality improvements, that would be a pretty big boon for contributors and maintainers.

cztomsik commented:
I know I'm late but the alternative api provided here is exactly what I want (and I'm considering either forking winit or just using glfw because it's possible to keep the control that way)

Events Loop 2.0 #459 seems to go in the opposite direction :-/

To give a bit more context, we run inside of node.js event loop which is on the main thread and we need to return if there were no events so the other work can be done (async I/O), we also want to switch between polling and blocking depending on whether we are animating or not to save power

Yes, it's a bit specific use-case but at the same time, I believe this would be the same for python and maybe even other platforms

mitchmindtree commented:
@cztomsik the ability to return from the main loop in EventLoop 2.0 is also something I'm interested in - it might be worth opening a new issue specifically for this.

vberger commented:
Event Loop 2.0 does introduce a `run_return` method as an extension trait for platforms that can support returning from the event loop though :

https://github.com/tomaka/winit/blob/eventloop-2.0/src/platform/desktop.rs

cztomsik commented:
`run_return` is very different thing - you're given the control back, but after the loop is already destroyed (or at least it was the case in the macos code I've seen before but I can't find it right now)

we will probably just migrate to glfw, because it's better fit for us (more explicit api in overall and for some reason even much less cpu usage), I only wanted to give you feedback about some specific but real-world usage

Osspial commented:
I'm closing this because of EventLoop 2.0.
